### PR TITLE
Sdesk 5581 add tags to asset metadata

### PR DESCRIPTION
**Pull request by dev-tsoni**

### CODE REVIEW

File: scripts/core/superdesk-api.d.ts [hunk 1]
```diff
@@ -1104,6 +1109,7 @@ declare module 'superdesk-api' {
         patchRaw<T extends IBaseRestApiResponse>(endpoint, id: T['_id'], etag: T['_etag'], patch: Partial<T>): Promise<T>;
         delete<T extends IBaseRestApiResponse>(endpoint, item: T): Promise<void>;
         uploadFileWithProgress<T>(endpoint: string, data: FormData, onProgress: (event: ProgressEvent) => void): Promise<T>;
+        searchTags(endpoint, params?: Dictionary<string, any>): Promise<IAutoTaggingSearchResult>;
```
Comment (MarkLark86) [inline scripts/core/superdesk-api.d.ts:1112]: This API entry is specific to `SAMS`, so it should belong in the `SAMS API` instead:
`/scripts/extensions/sams/src/api/assets.ts`

File: scripts/extensions/sams/src/components/assets/assetEditor.tsx [hunk 2]
```diff
@@ -1,6 +1,9 @@
 // External modules
 import * as React from 'react';
 import {connect} from 'react-redux';
+import {OrderedMap} from 'immutable';
```
Comment (MarkLark86) [inline scripts/extensions/sams/src/components/assets/assetEditor.tsx:4]: Why use `OrderedMap` instead of an `Array`

File: scripts/extensions/sams/src/components/assets/assetEditor.tsx [hunk 3]
```diff
@@ -24,11 +27,34 @@ interface IProps {
     sets: Array<ISetItem>;
     fields?: Array<keyof IAssetItem>;
 }
+interface ITagUi {
```
Comment (MarkLark86) [inline scripts/extensions/sams/src/components/assets/assetEditor.tsx:30]: You could use `IAssetItem['tags']` here instead of re-declaring the same interface

File: scripts/extensions/sams/src/components/assets/assetEditor.tsx [hunk 4]
```diff
@@ -141,6 +168,41 @@ class AssetEditorComponent extends React.PureComponent<IProps> {
                         </FormRow>
                     </FormGroup>
                 )}
+                {this.fieldEnabled('tags') && (
+                    <FormGroup>
+                        <FormRow>
+                            <Autocomplete
+                                key={nextId.name}
```
Comment (MarkLark86) [inline scripts/extensions/sams/src/components/assets/assetEditor.tsx:175]: This `key` needs to change with every update, so that the search box's text is cleared after a user selects a Tag to be added

File: scripts/extensions/sams/src/components/assets/assetEditor.tsx [hunk 5]
```diff
@@ -141,6 +168,41 @@ class AssetEditorComponent extends React.PureComponent<IProps> {
...
+                                label={gettext('Tags')}
+                                value={''}
+                                keyValue="name"
+                                items={[]}
+                                search={(searchString, callback) => {
+                                    let cancelled = false;
+
+                                    superdeskApi.dataApi.searchTags('sams/assets/tags', {
```
Comment (MarkLark86) [inline scripts/extensions/sams/src/components/assets/assetEditor.tsx:183]: This would become `samsApi.assets.searchTags(searchString + '*')

File: scripts/extensions/sams/src/components/assets/uploadAssetModal.tsx [hunk 6]
```diff
@@ -128,7 +128,11 @@ export class UploadAssetModalComponent extends React.Component<IUploadAssetModal
 
         for (field in asset) {
             if (['_id', 'length'].includes(field) === false) {
-                data.append(field, asset[field] as string);
+                if (typeof asset[field] === 'object') {
+                    data.append(field, JSON.stringify([asset[field]]));
```
Comment (MarkLark86) [inline scripts/extensions/sams/src/components/assets/uploadAssetModal.tsx:132]: Why is the value being supplied as an array? i.e.
```typescript
JSON.stringify([asset[field]])
```
vs
```typescript
JSON.stringify(asset[field])
```

The value being sent to the server should be the same value as stored in the local Asset, in a jsonified string

File: scripts/extensions/sams/src/components/assets/assetEditor.tsx (hunk 2 above)
Comment (dev-tsoni) [reply scripts/extensions/sams/src/components/assets/assetEditor.tsx:4]: It is being converted into array below, where this function is called, line 188.

Comment (dev-tsoni) [reply scripts/extensions/sams/src/components/assets/assetEditor.tsx:4]:
```
description: ""
filename: "City7.jpg"
length: 1234818
mimetype: "image/jpeg"
name: "City7.jpg"
set_id: "602d04c5ee05b7bb601ebabf"
state: "draft"
tags: {name: "item6", code: "item6"}
```
This is the response without [ ].

File: scripts/extensions/sams/src/components/assets/assetEditorPanel.tsx [hunk 7]
```diff
@@ -69,16 +71,59 @@ export class AssetEditorPanelComponent extends React.PureComponent<IProps, IStat
...
 
     onChange<K extends keyof IAssetItem>(field: K, value: IAssetItem[K]) {
-        this.setState((prevState: IState) => ({
-            updates: {
-                ...prevState.updates,
-                [field]: value,
-            },
-            isDirty: true,
-        }));
+        if (field === 'tags') {
```
Comment (MarkLark86) [inline scripts/extensions/sams/src/components/assets/assetEditorPanel.tsx:79]: This would be better moved into the `AssetEditor` itself, otherwise this code (and `addTag`, `removeTag`) would need to be duplicated to `UploadAssetModal` and `EditAttachmentModal` as well

File: scripts/extensions/sams/src/components/assets/assetEditorPanel.tsx [hunk 8]
```diff
@@ -149,12 +194,27 @@ export class AssetEditorPanelComponent extends React.PureComponent<IProps, IStat
...
                             ]}
                         />
+                        <FormGroup>
+                            <FormRow>
+                                {this.state.updates?.tags?.map((tag) => (
+                                    <Tag
+                                        key={this.state.updates?.tags?.indexOf(tag)}
+                                        text={tag.name}
+                                        onClick={() => {
+                                            this.removeTag(this.state.updates?.tags?.indexOf(tag)!);
```
Comment (MarkLark86) [inline scripts/extensions/sams/src/components/assets/assetEditorPanel.tsx:212]: Again, this would be better off as a function in this class/component, and in turn that function will call `this.props.onChange` with the new tag value

File: scripts/extensions/sams/src/components/assets/assetEditor.tsx (hunk 2 above)
Comment (MarkLark86) [reply scripts/extensions/sams/src/components/assets/assetEditor.tsx:4]: I'm still confused here, `tags` should always be an Array.

File: scripts/extensions/sams/src/components/assets/uploadAssetModal.tsx (hunk 6 above)
Comment (MarkLark86) [reply scripts/extensions/sams/src/components/assets/uploadAssetModal.tsx:132]: This code is very specific to `tags`. If the value is an `object`, `JSON.stringify` should suffice for all data types. `tags` should always be an array, so it should work here without encasing in `[` and `]`

File: scripts/extensions/sams/src/components/assets/assetEditorPanel.tsx (hunk 8 above)
Comment (dev-tsoni) [reply scripts/extensions/sams/src/components/assets/assetEditorPanel.tsx:212]: In onChange,we pass the tag, but can not pass the operation to be performed. I'll have to pass a string with it that will specify that the tag is needed to be added or removed. I'll be adding the code in commit though.

File: scripts/extensions/sams/src/components/assets/assetEditorPanel.tsx (hunk 7 above)
Comment (dev-tsoni) [reply scripts/extensions/sams/src/components/assets/assetEditorPanel.tsx:79]: `assetEditorPanel`, `uploadAssetModal` and `editAttachmentModal`, all have their own custom `onChange` and states, I think I'll have to probably add `addTag` and `removeTag` separately to them all, customised according to their states.

File: scripts/extensions/sams/src/components/assets/uploadAssetModal.tsx [hunk 9]
```diff
@@ -186,7 +190,7 @@ export class UploadAssetModalComponent extends React.Component<IUploadAssetModal
 
         return (
             <FileUploadModal
-                dark={true}
+                dark={false}
```
Comment (MarkLark86) [inline scripts/extensions/sams/src/components/assets/uploadAssetModal.tsx:193]: Why was this changed from dark to light?

Comment (dev-tsoni) [reply scripts/extensions/sams/src/components/assets/uploadAssetModal.tsx:193]: The tags were not being visible in dark.

File: scripts/extensions/sams/src/components/assets/assetEditor.tsx [hunk 10]
```diff
@@ -20,27 +22,119 @@ import {getHumanReadableFileSize} from '../../utils/ui';
 interface IProps {
     asset: Partial<IAssetItem>;
     disabled?: boolean;
-    onChange(field: string, value: string): void;
+    onChange/* <K extends keyof IAssetItem> */(field: string, value: any): void;
```
Comment (MarkLark86) [inline scripts/extensions/sams/src/components/assets/assetEditor.tsx:25]: What's wrong with the types on this function prop? i.e. can you un-comment so it's like this:
```typescript
onChange<K extends keyof IAssetItem>(field: K, value: IAssetItem[K]): void;
```

Comment (dev-tsoni) [reply scripts/extensions/sams/src/components/assets/assetEditor.tsx:25]: Ohh, my bad. removing it.

**Review by MarkLark86 — APPROVED** [review]
